lyncs_io/dask_io.py

```diff
@@ -0,0 +1,188 @@
+        """
+        # pylint: disable=C0415
+        import dask
+
+        return dask
+
+    def __init__(self, filename):
+
+        self.filename = filename
+
+    def load(self, domain, dtype, header_offset, chunks=None, order="C"):
+        """
+        Reads the global domain from a file and loads it in a dask array
+
+        Parameters
+        ----------
+        domain : list
+            Global data domain.
+        dtype: data-type
+            dask/numpy data-type for the array
+        order: str
+            whether data are stored in row/column
+            major ('C', 'F') order in memory
+        header_offset: int
+            offset in bytes to where the
+            data start in the file.
+
+        Returns:
+        --------
+        array : dask array
+            A lazy evaluated array to be computed on demand
+        """
+        array = numpy.memmap(
+            self.filename,
+            mode="r",
+            shape=domain,
+            dtype=dtype,
+            offset=header_offset,
+            order=order,
+        )
+
+        return self.dask.array.from_array(array, chunks=chunks)
+
+    def save(self, array):
+        """
+        Writes the array in a npy file in parallel using dask
+
+        Parameters
+        ----------
+        array : dask/numpy array
+            The array to be written
+
+        Returns:
+        --------
+        array : dask array
+            A lazy evaluated array to be computed on demand
+        """
+
+        header = _build_header_from_dask_array(array)
+
+        return self.dask.array.map_blocks(
+            _write_blockwise_to_npy,
+            array,
+            self.filename,
+            header,
+            array.shape,
+            chunks=array.chunks,
+            dtype=array.dtype,
+        )
+
+
+def _get_dask_array_header_offset(header):
+    stream = BytesIO()
+    numpy.lib.format._write_array_header(stream, header)
+
+    return len(stream.getvalue())
+
+
+def _build_header_from_dask_array(array, order="C"):
+    header_dict = {"shape": array.shape}
+    header_dict["fortran_order"] = bool(order == "F")
+    header_dict["descr"] = numpy.lib.format.dtype_to_descr(array.dtype)
+
+    return header_dict
+
+
+def _build_header_from_file(filename):
+
+    with open(filename, "rb") as fptr:
+        version = numpy.lib.format.read_magic(fptr)
+        numpy.lib.format._check_version(version)
+        shape, fortran_order, dtype = numpy.lib.format._read_array_header(fptr, version)
+
+        header_dict = {"shape": shape}
+        header_dict["fortran_order"] = fortran_order
+        header_dict["descr"] = dtype
+
+    return header_dict
+
+
+def _write_npy_header(filename, header):
+
+    lock_path = filename + ".lock"
+
+    with FileLock(lock_path):
+        write_header = True
+        if os.path.exists(filename):
+            if header == _build_header_from_file(filename):
+                write_header = False
```

Comment:
This looks fine, but not exactly what we discussed yesterday.
In this way the reading of the header is serialized, since all processed have to acquire the lock.
This is bad for performance... 
The reading of the header should be done without lock and if it matches they proceed to writing the data.
Only for the writing the header one process has to acquire the lock and write.. and if there is a lock active all the others needs to wait before proceeding to reading the header and make sure it is correct.. Something like that

Reply:
I agree. However, evaluating if the file exists or not outside the lock wouldn't create a possibility for a race condition?

Reply:
The filesystem should take care of that. It is important to enable the lock for writing the header, so only one does it, and do not read the file if the lock is active. But I understand it is tricky...

Reply:
@sbacchio made the changes. If there is nothing further on the header writing it should be good to go

Reply:
@cstyl Added a comment on it. When it is resolver it can be merged